### ISSUE TITLE
Use a single "if" statement for "if (firstCheck)"

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/angular#712]
Example: https://github.com/dart-lang/angular/blob/089decdc8d646f0601a98563525081f54fefd00d/_goldens/test_files/empty_properties.template_release.golden#L100-L118

I naively thought dart2js would optimize this, but it doesn't seem to. Test case:

```dart
void main(dynamic foo) {
  detectChangesInternal(foo);
}

void detectChangesInternal(int cdState) {
  bool firstCheck = (cdState == 0);
  if (firstCheck) {
    print(1);
  }
  if (firstCheck) {
    print(2);
  }
  if (firstCheck) {
    print(3);
  }
  if (firstCheck) {
    print(4);
  }
}
```

... emits ...

```js

main: [function(foo) {
  var firstCheck = J.$eq$(foo, 0);
  if (firstCheck)
    P.print(1);
  if (firstCheck)
    P.print(2);
  if (firstCheck)
    P.print(3);
  if (firstCheck)
    P.print(4);
}, "call$1", "main__main$closure", 2, 0, 7]
```

**Comment by matanlurey** [comment]
Note that making `firstCheck` `final` doesn't seem to change the output.

**Comment by matanlurey** [comment]
Filed https://github.com/dart-lang/sdk/issues/31462.

**Comment by alorenzen** [comment]
Still worth tracking here?

**Comment by matanlurey** [comment]
Yes. Its unlikely Dart2JS will be help able to help us in the short-term.

**Comment by matanlurey** [comment]
For anyone wondering if this is simple, it is not. You need to deal with a lot of:

```dart
  if (dynamicPropertiesMethod.isNotEmpty) {
    targetMethod.addStmts(dynamicPropertiesMethod.finish());
  }
```

... in 10+ places in the view compiler. 

One option would be doing something like:

```dart
class CompileMethod {
    // Existing code.
    void addStmts(List<o.Statement> stmts) {
      if (stmts.isEmpty) return;
      _bodyStatements.addAll(stmts);
    }

    // New code.
    void addStmtsFirstCheck(List<o.Statement> stmts) {
      // If the last stmt is _also_ a first check stmt, just add to that if block instead.
      // Else call addStmts.
    }
}
```

**Comment by matanlurey** [comment]
There is one case that is extra hard, which is `!throwOnChanges && firstCheck`.

We don't know, at compile-time, whether it is safe to collapse that into `if (firstCheck)`. We could add yet another method that tries to embed `if (!throwOnChanges) { ... }` into the `if (firstCheck) { ... }` block, but that means adding yet-another change detection specific `addStmts`.

(Not the worst thing to do, but I wouldn't do it in the same CL)

**Comment by matanlurey** [comment]
Upon reflection, https://github.com/dart-lang/angular/issues/712#issuecomment-403189729 is a pretty rare case - it is used in lifecycle events only, which won't usually (or ever) be immediately after another `if (firstCheck) { ... }`, so I think we can skip this case entirely.

**Comment by matanlurey** [comment]
Fixed in #1461.
